sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp

```diff
@@ -61,122 +67,269 @@ void RingOutlierFilterComponent::faster_filter(
+      float azimuth_diff = curr_azimuth - prev_azimuth;
+      azimuth_diff = azimuth_diff < 0.f ? azimuth_diff + 36000.f : azimuth_diff;
+      return std::max(curr_distance, prev_distance) <
+               std::min(curr_distance, prev_distance) * distance_ratio_ &&
+             azimuth_diff < 100.f;
+    };
+
+  // tmp vectors to keep track of walk/ring state while processing points in order (cache efficient)
+  std::vector<WalkInfo> walks;                // all walks
+  std::vector<RingWalkInfo> rings_curr_walk;  // each ring curr walk info
+  std::vector<size_t> rings_first_walk_idx;   // each ring very first walk idx
+  std::vector<size_t> points_walk_idx;        // each point walk idx
+
+  walks.reserve(max_rings_num_);
+  // initialize each ring with empty walk
+  rings_curr_walk.resize(max_rings_num_, RingWalkInfo{-1U, 0., 0.});
+  rings_first_walk_idx.resize(max_rings_num_, -1U);
+  // points are initially associated to no walk
+  points_walk_idx.resize(input->width * input->height, -1U);
+
+  int invalid_ring_count = 0;
+
+  // Build walks and classify points
+  for (const auto & [raw_p, point_walk_idx] :
+       ranges::views::zip(input->data | ranges::views::chunk(input->point_step), points_walk_idx)) {
+    const uint16_t ring_idx = getPointRing(raw_p.data());
+    const float curr_azimuth = getPointAzimuth(raw_p.data());
+    const float curr_distance = getPointDistance(raw_p.data());
+
+    if (ring_idx >= max_rings_num_) {
+      // Either the data is corrupted or max_rings_num_ is not set correctly
+      // Note: point_walk_idx == -1 so the point will be filtered out
+      invalid_ring_count += 1;
+      continue;
+    }
 
-    walk_first_idx = 0;
-    walk_last_idx = -1;
+    auto & ring = rings_curr_walk[ring_idx];
+    if (ring.walk_idx == -1U) {
+      // first walk ever for this ring
+      walks.push_back(WalkInfo{1, curr_distance, curr_azimuth, curr_distance, curr_azimuth});
+      const auto walk_idx = walks.size() - 1;
+      point_walk_idx = walk_idx;
+      rings_first_walk_idx[ring_idx] = walk_idx;
+      ring.walk_idx = walk_idx;
+      ring.prev_azimuth = curr_azimuth;
+      ring.prev_distance = curr_distance;
+      continue;
+    }
 
-    for (size_t idx = 0U; idx < indices.size() - 1; ++idx) {
-      const size_t & current_data_idx = indices[idx];
-      const size_t & next_data_idx = indices[idx + 1];
-      walk_last_idx = idx;
+    auto & walk = walks[ring.walk_idx];
+    if (isSameWalk(curr_distance, curr_azimuth, ring.prev_distance, ring.prev_azimuth)) {
+      // current point is part of previous walk
+      walk.num_points += 1;
+      walk.last_point_distance = curr_distance;
+      walk.last_point_azimuth = curr_azimuth;
+      point_walk_idx = ring.walk_idx;
+    } else {
+      // previous walk is finished, start a new one
+      walks.push_back(WalkInfo{1, curr_distance, curr_azimuth, curr_distance, curr_azimuth});
```

Comment:
We reserve `max_rings_num_` walks in the `walks` vector, so it needs to be resized multiple times when appending on the order of 16000 walks, as is the case in the "many walks" section.

Reply:
Indeed, I only allocated the strict minimum (1 walk per ring), but a bigger value might be wiser.